File: ruby/lib/EnemyStarShip.rb
require_relative 'ShotResult'
require_relative 'EnemyToUI'

module Deepspace
  class EnemyStarShip
    
    attr_reader :ammoPower
    attr_reader :damage
    attr_reader :loot
    attr_reader :name
    attr_reader :shieldPower
    
    
    public

    def initialize(name, ammoPower, shieldPower, loot, damage)
      @ammoPower = ammoPower
      @name = name
      @shieldPower = shieldPower
      @loot = loot
      @damage = damage
    end
    
    def self.newCopy(e)
      EnemyStarShip.new(e.name, e.ammoPower, e.shieldPower, e.loot, e.damage)
    end

    def fire
      @ammoPower
    end
    
    def protection
      @shieldPower
    end

    def receiveShot(shot)
      if protection < shot
        return ShotResult::DONOTRESIST
      else  
        return ShotResult::RESIST
      end
    end
    
    def to_s
      "Name: #{@name}; Damage: #{@damage}; Loot: #{@loot}; AmmoPower: #{@ammoPower}; ShieldPower: #{@shieldPower} "
    end
    
    def getUIversion
      EnemyToUI.new(self)
    end
    
  end
end

File: java/src/deepspace/EnemyStarShip.java
package deepspace;

class EnemyStarShip implements SpaceFighter {

    private float ammoPower;
    private float shieldPower;
    private String name;
    private Loot loot;
    private Damage damage;
    
    
    EnemyStarShip(String n, float a, float s, Loot l, Damage d){
	name = n;
	ammoPower = a;
	shieldPower = s;
	loot = new Loot(l.getNSupplies(), l.getNWeapons(), l.getNShields(), l.getNHangars(), l.getNMedals());
        if (d instanceof NumericDamage){
            damage = new NumericDamage((NumericDamage)d);
        }else{
            damage = new SpecificDamage((SpecificDamage)d);
        }
    }

    EnemyStarShip(EnemyStarShip e){
	this(e.getName(), e.getAmmoPower(), e.getShieldPower(), e.getLoot(), e.getDamage());
    }

    EnemyToUI getUIversion(){
	return new EnemyToUI(this);
    }

    @Override
    public float fire(){
	return ammoPower;
    }
    
    public float getAmmoPower(){
	return ammoPower;
    }

    public Damage getDamage(){
	return damage;
    }

    public Loot getLoot(){
	return loot;
    }

    public String getName(){
	return name;
    }
    
    public float getShieldPower(){
	return shieldPower;
    }
    
    @Override
    public float protection(){
	return shieldPower;
    }
    
    @Override
    public ShotResult receiveShot(float shot){
	if (protection() < shot){
	    return ShotResult.DONOTRESIST;
	}
	else{
	    return ShotResult.RESIST;
	}
    }

}
